### PR TITLE
docs: workflow Spring SDK

### PR DESCRIPTION
**Pull request by aludwiko**

### CODE REVIEW

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:41]:
```suggestion
   * Returns the initial empty state object. This object will be passed into the
```

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:99]:
```suggestion
   * @return A workflow definition in a form of steps and transitions between them.
```

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:129]: I like the example.

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:193]:
```suggestion
     * A workflow effect type that contains information about the transition to the next step. This could be also a special transition to pause or end the workflow.
```

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:340]: Just to be sure. It's a Step builder but the class name is a `CallBuilder`?

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:378]:
```suggestion
       * Transition to the next step based on the result of the step action.
```
Maybe?

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:381]: `???` is missing?

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:406]: `???` is missing?

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (franciscolopezsancho) [reply sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:378]: If you agree feel free to change the same one below.

**Review by franciscolopezsancho — COMMENTED** [review]
Looking good. My only real question is the `???`. The rest feel free to skip it.

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (aludwiko) [reply sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:340]: Renamed

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (aludwiko) [reply sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:378]: thanks

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (aludwiko) [reply sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java]: yeah, sorry, a work in progress, I should put this PR as a draft, but it is fixed now

File: docs/src/modules/spring/pages/workflow-entities.adoc
Comment (octonato) [inline docs/src/modules/spring/pages/workflow-entities.adoc:9]: I don't think we should be prescriptive about what should be put on the Workflow state. I mean, we don't need to say something like 'it should or it should not'. 

Better if we say that a Workflow has a state and that on each iteration we can update it. We can use it to track progress or to validate steps. Or whatever we see fitting.

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java:111]: I think this should be an `error`. Not essential to explain workflows though. 

Also, I think we should start with guard clauses and then the meat, for example:

```java
if (transfer.amount() <= 0) {
  return effects().error("transfer amount should be greater than zero");
} else if (currentState() != null) {
  return effects().error("transfer already started"));
} else {
  //... start the workflow now
}

```

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java:108]: What about this?

```java
return effects()
    .updateState(new TransferState(transfer)) 
    .transition(new Withdraw(transfer.from(), transfer.amount()), "withdraw") 
    .thenReply(new Message("transfer started")); 
```

For documentation, is nice if we have it more compact. 

Also nice to have it more self-contained. So, "withdraw" instead of `withdrawStepName`. The rational here is that, if we use a variable, we need to explain it. We need to first show that we defined it and then use it. 

It's certainly a best-practice, but we can leave this to the user since it's a general best-practice. IMO, we should only suggest best-practice in domains that may not be familiar for the user. For general stuff, better if let the users use their own judgement. It saves of time and make it easier to document.

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferState.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferState.java:12]: I would reduce the status. 

To demo `pause` we can require a second `sign-off`. The state can have the name of the initiator and then the name of the second person that authorizes the transfer, for instance.

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferState.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferState.java:21]: For documentation page, we can make it more explicit, ie: `withStatus(ACCEPTED)`. 

Then we can show the `withStatus` method. If we have one specific method each time, we won't show all the methods impl because that's too long. 

When I see the following in the docs
```java
TransferState successfulWithdrawal = transferState.asSuccessfulWithdrawal();
```
I don't know what `asSuccessfulWithdrawal` does unless I have seen the impl for `asSuccessfulWithdrawal` before.

While if I have seen the `withStatus` impl before and later I see

```java
TransferState successfulWithdrawal = transferState.withStatus(SUCCESSFUL_WITHDRAWAL);
```

it becomes immediately clear what is happening.

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java:65]: Maybe better to use lambdas to make it self-contained. 

Also, the callouts numbers will be in order.

File: samples/spring-transfer-workflow/src/main/java/com/example/wallet/Balance.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/wallet/Balance.java:3]: We can add the domain types inside the entities. Same story, all self-contained. 

Make it also easier to document.

File: samples/spring-transfer-workflow/src/main/java/com/example/wallet/WalletEntity.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/wallet/WalletEntity.java:29]: heads-up:

Nice to have logging in the samples, but let's make sure that it doesn't appears in the docs.

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java
Comment (octonato) [inline samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java:64]: This long calls are forcing a horizontal scroll. Gets more problematic if we have a callout at the end of the line.

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java
Comment (octonato) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java:356]: Nitpicking Alert
Conventionally, the `@return` is the last in the list.

**Review by octonato — COMMENTED** [review]
Looking good. 

As said in Slack, I think we should have a simple `pause` example. Like extra sign-off. Could be conditional: requires sign-off if amount is > 1000

We can also simplify the status

* PENDING - waiting for sign-off
* STARTED
* SUCCESSFUL_WITHDRAWAL
* COMPLETED

So no ACCEPTED or REJECTED. It either goes straight to STARTED or it waits for a sign-off.

File: samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java
Comment (aludwiko) [reply samples/spring-transfer-workflow/src/main/java/com/example/transfer/TransferWorkflowEntity.java]: how about now?

**Comment by aludwiko** [comment]
ready for review

File: docs/src/modules/spring/pages/workflow-entities.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/workflow-entities.adoc:21]:
```suggestion
Now we can focus on the workflow implementation itself. A workflow has state - which can be updated in command handlers - and transition methods. During state modeling we should consider what information is required for validation, running the steps, collecting data from steps or tracking the workflow progress.
```

File: samples/spring-transfer-workflow/src/main/resources/logback.xml
Comment (efgpinto) [inline samples/spring-transfer-workflow/src/main/resources/logback.xml:37]: might be useful to make sure we have the [new structure](https://github.com/lightbend/kalix-jvm-sdk/pull/1451) in place for this new sample.. but I guess you can that that now or later, whatever you prefer.. just trying to ensure we don't forget about it

**Review by efgpinto — APPROVED** [review]
LGTM with some small suggestions documentation wise.

File: docs/src/modules/spring/pages/workflow-entities.adoc
Comment (aludwiko) [reply docs/src/modules/spring/pages/workflow-entities.adoc:79]: yeah, next time :)

File: samples/spring-transfer-workflow/src/main/resources/logback.xml
Comment (aludwiko) [reply samples/spring-transfer-workflow/src/main/resources/logback.xml:37]: yeah, I forgot to push it

File: docs/src/modules/spring/pages/workflow-entities.adoc
Comment (octonato) [inline docs/src/modules/spring/pages/workflow-entities.adoc:37]: shall we also mention: "or retrieve its current state"?

**Review by octonato — APPROVED** [review]
Looking great! 

I left a minor suggestion, but we can merge as is.

**Pull request merged**
